Signal Forms Part 3 (#41)
angular-buch/website-articles: blog/2025-10-signal-forms-part3/README.md

---
title: 'Angular Signal Forms Part 3: Child Forms and Custom UI Controls'
author: Danny Koppenhagen and Ferdinand Malcher
mail: dannyferdigravatar@fmalcher.de # Gravatar
published: 2025-10-27

> lastModified: 2025-10-27

Reviewer: ```suggestion
published: 2025-10-21
lastModified: 2025-10-21
```

```html
<form (submit)="submit($event)">
  <!-- ... -->
  <app-identity-form [identity]="registrationForm.identity" />
  <!-- ... -->
</form>
```

The child form now seamlessly integrates with the parent form.
`RegistrationForm` as the main form component puts all things together: It holds the complete data model, applies the schemas (including the child schema), and manages form submission and validation.
Parts of the form are passed to sub components like `IdentityForm`, which bind to the fields and handle their own UI logic.

## Creating custom Form UI Controls

So far, we've used standard HTML form elements like `<input>`, `<select>`, and `<textarea>` to build our forms.
However, in real-world applications, we often need form controls that go beyond standard HTML input elements.
Think of a date picker, a rich text editor, a multi-select dropdown, a counter control, wrappers for third-party component libraries, or custom UI elemnts that fit specific design requirements.

Signal Forms provide an interface that allows you to create custom form components that integrate seamlessly with the Signal Forms ecosystem.

> Our goal is to create custom component that can be used just like native HTML form elements with the `Field` directive.

Reviewer: ```suggestion
Our goal is to create a custom component that can be used just like native HTML form elements with the `Field` directive.
```

The inputs `errors` and `disabled` are optional: These values are automatically provided by the `Field` directive when the control has errors or is disabled.
We can use the `disabled` signal to clear the value when the control is disabled by using an `effect`.

Our component can define additional inputs as needed:
The inputs `label` and `selectOptions` receive the label for the form control and the list of all options to select from.

```typescript
import { Component, effect, input, model } from '@angular/core';
import { FormValueControl, ValidationError, WithOptionalField } from '@angular/forms/signals';

@Component({ /* ... */ })
export class Multiselect implements FormValueControl<string[]> {
  readonly value = model<string[]>([]);
  readonly errors = input<readonly WithOptionalField<ValidationError>[]>([]);
  readonly disabled = input<boolean>(false);

  readonly selectOptions = input.required<string[]>();
  readonly label = input.required<string>();

  constructor() {
    effect(() => {
      if (this.disabled()) {
        this.value.set([]);
      }
    });
  }
}
```

Next, we implement the template where we use native HTML elements for accessibility.

> We want to keep this example as simple as possible, so please note that this is not a production-ready multiselect component.

Author: Das ist ein super Beispiel, aber auch nicht mehr als das: Tastaturnavigation mit den Pfeiltasten geht nicht, man kann es nicht mit *Escape* schließen, die Reihenfolge der ausgewählten Optionen entspricht nicht der ursprünglichen Reihenfolge (sondern der ausgewählten).

Ich möchte gern vermeiden, dass jemand das 1:1 kopiert und ins Projekt übernimmt – denn es geht in dem Blogpost ja gar nicht um das Multiselect, sondern darum, wie man ein eigenes Control mit dem Interface baut.

Wir könnten das ja etwas umdrehen und sagen:

> We want to keep the example as simple as possible, so please remember that for a production-ready multiselect component we need to address a few more details, e.g. proper keyboard navigation.

> We want to keep this example as simple as possible, so please note that this is not a production-ready multiselect component.

Reviewer: ```suggestion
We want to keep the example as simple as possible, so please remember that for a production-ready multiselect component we need to address a few more details, e.g. proper keyboard navigation.
```
